src/core/primitives.rs

```diff
@@ -451,3 +438,53 @@ fn refresh_coordinate_keys(
         })
         .collect::<RevisionVec<_, _>>()
 }
+
+/// Attempts opening the Covercrypt encapsulation using the given USK. Returns
+/// the encapsulated key and associated rights upon success, otherwise returns `None`.
+pub fn full_decaps(
+    encapsulation: &XEnc,
+    msk: &MasterSecretKey,
+) -> Result<Vec<(Right, Secret<SHARED_SECRET_LENGTH>)>, Error> {
+    // A = ⊙ _i (α_i. c_i)
+    let A = encapsulation.c.first().unwrap() * &(&msk.tsk.s / &msk.tsk.tracers.front().unwrap().0);
+    let mut rights_list: Vec<(Right, Secret<SHARED_SECRET_LENGTH>)> = Vec::new();
+    for enc in &encapsulation.encapsulations {
+        for (right, key) in msk.secrets.iter() {
+            for k in key {
+                let list: Vec<Right> = rights_list.clone().into_iter().map(|r| r.0).collect();
+                if list.contains(right) {
+                    break;
+                } else {
+                    if let Some(S) = S(&k.1, enc, A.clone()) {
+                        let (tag, ss) =
+                            j_hash(&S, &encapsulation.c, &encapsulation.encapsulations)?;
+                        if tag == encapsulation.tag {
+                            rights_list.push((right.clone(), ss));
+                        }
+                    } else {
+                        continue;
+                    }
+                }
+            }
+        }
+    }
+    Ok(rights_list)
+}
+
+fn S(key: &RightSecretKey, enc: &Encapsulation, A: R25519CurvePoint) -> Option<[u8; 32]> {
+    match (key, enc) {
+        (RightSecretKey::Hybridized { sk, dk }, Encapsulation::Hybridized { E, F }) => {
+            let mut K1 = h_hash(R25519::session_key(&sk, &A).unwrap());
+            let K2 = MlKem512::dec(&dk, &E).unwrap();
+            let S = xor_3(&F, &K1, &K2);
```

Comment:
This should be zeroized I think. Maybe modify the XORing functions to take a `&mut [u8; 32]` instead of returning a `[u8; 32]` and deref mut on a secret.